### ISSUE TITLE
All Instruments: can run appropriate existing system tests and IOC tests

### ISSUE THREAD

**Issue by ChrisM-S** [ISISComputingGroup/IBEX#5490]
As a system developer and tester I would like to be able confirm that changes to an instrument system, for example a change to Windows 10 (see #5489), or software install or hardware update, have left the system in a _demonstrably_ working state.

Some IOC test framework and System tests in particular (run in groups or individually) could go a long way to establishing that no regression in overall system functionality has occurred, regardless of the system change.  This could be used optionally to test the validity of a IBEX deployment - but the main goal is the wider system (hardware, platform, network, software, including IBEX).

***Acceptance Criteria***
  - [ ] Initially, a useful subset of IOC or system tests can be run on the target system, individually or grouped (more tests should be easily be added/converted easily once this ticket has been done)
   - [ ] The tests chosen are appropriate to "proving/qualifying" the new system so should ideally challenge the system.
  - [ ] Tests included take account of comments in the notes below.
  - [ ] The use of these tests on a live system in a diagnostic capacity should be considered.

***Notes***
1. It might be most useful to consider starting with the system to test "as-is" rather than always generating extra test artefacts (new clean folders etc).
2. The systems (on VHDs) could be snapshotted first, tested and then restored to a previous state to guarantee to clean up tests safely - this could also remove some need for some test clean-up code.  It is important that any modifications for testing can be _completely_ removed from a _live_ instrument under test.
3. Hardware "in the loop" could and should be tested where possible so the ability to switch off simulation and drive hardware safely but diagnostically on a test or IOC basis will greatly expand the value of the testing.

**Comment by John-Holt-Tessella** [comment]
While I understand the reason for this ticket I do not agree with the implementation. System tests are tests which can assert something about an instrument. Running them on an actual instrument is bad. 
I think what you really want is initially a checklist of things and then we can develop that into non-destructive system tests.

**Comment by kjwoodsISIS** [comment]
I agree with John.  System tests are not appropriate for a production system, not least because the system tests require a supporting infrastructure which we would never install on an instrument PC.
Furthermore, if we are having to modify an instrument PC in order to run the tests, then we are not  testing IBEX on the genuine target.

Also, I am not clear about the scope here.  The description says "the main goal is the wider system (hardware, platform, network, software, including IBEX)".  But the acceptance criteria say "a useful set of IOC or system tests" (which I take to mean the IBEX system tests).  Where are the tests that ensure a change to Windows 10 has not broken the hardware, platform, network and other software elements?  I think the acceptance criteria need to be specific - performing action X should generate in end-result Y.

Ultimately, I think what we need is something more akin to "Acceptance Tests", which you run after you have completed all your system tests and are confident that your system will run correctly in a production environment.  This is something that scientists do (to varying extents) when we deploy a new release of IBEX - they run a few of their favourite scripts or perform a few actions with the GUI to satisfy themselves that everything works as expected.  We could do the same.  Indeed, we could work with the scientists to create and maintain a set of "acceptance tests".  However, we should not underestimate the scale of such a task.  Every instrument will have its own unique set of "acceptance tests", because every instrument is unique.  But we can start small and grow incrementally.

**Comment by ChrisM-S** [comment]
Hmm, maybe need a rethink!

Would inverting things such that we ran system and IOC tests on exactly the same hardware, VM, disk technology, network connection and OS, patch level and virus checker as the identical real instrument target configuration be likely to spot any regression in any component from the development system? if so could we re-use the existing testing work we've done in the wider context of proving the whole system?

Could simple (new) acceptance tests be automated as a small set of heuristics which give confidence though an argument of correctness (e.g. we begin a run on a DAEIII, save a data file and check that the file contains plausible zero data of the right size and with at least a few non-zero parameters read from the DAE - to test all these paths).  Adding a logging block to a temperature controller on RS232 and a motion device on a private network.  Again just checking the data file contained values which means that everything from code to device to network functioned as expected might be sufficient?
